### PR TITLE
[WIP] AMO client sync

### PR DESCRIPTION
**Pull request by Natim**
Work in progress.

### CODE REVIEW

File: services/common/KintoAddonPluginBlocklist.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  /**
+   * Synchronize from Kinto server, if necessary.
+   *
+   * @param {int}  lastModified   the current date return by server.
+   * @param {Date} serverTime   the current date return by server.
+   */
+  maybeSync(lastModified, serverTime) {
+    let db = kintoClient();
+    let collection = db.collection(this.collectionName);
+    const self = this;
```
Comment (n1k0) [inline services/common/KintoAddonPluginBlocklist.js]: Couldn't we bind the generator function to `this` instead?

File: toolkit/mozapps/extensions/nsBlocklistService.js [hunk 2]
```diff
@@ -689,85 +696,62 @@ Blocklist.prototype = {
...
+
+      let text = this._loadBlocklistFromFile(jsonFile);
+      if (!text) {
+        LOG(`Blocklist::_loadBlocklist: no content found in ${filename}`);
+        return;
+      }
+
+      let parsedEntries;
+      try {
+        parsedEntries = JSON.parse("" + text);
```
Comment (leplatrem) [inline toolkit/mozapps/extensions/nsBlocklistService.js]: cast to string ?

**Comment by leplatrem** [comment]
- [x] Load JSON synchronously from disk
- [x] Preload JSON asynchronously (ex. session restore)
- [x] Write JSON on sync from `AddonPluginClient`
- [x] Plug `AddonPluginClient` with `nsBlocklistService` using a message
- [x] Update addons state when list is synced
- [x] Refactor `CertBlocklistClient` to share code with `AddonPluginClient`
- [ ] Make sure we use appropriate way of doing things vs. legacy (eg. `OS.Constants.Path.profileDir`)
- [ ] Optimize reload (?) : reload addons or plugins instead of boths each time one of them changes
- [x] Add missing test for appDir/profDir (/cc @tarekziade)
- [x] Plug GFX with JSON
- [ ] Strip text values from JSON when parsing

**Questions**
- Replace Kinto/SQLite with plain JSON / memory in `maybeSync()` ? → No. Keep kinto protocol abstract.
- Replace SQLite adapter with plain JSON manipulation ? → Looks like code complexity.
- Only one pref for `serverTime` instead of one per collection (cf. `PREF_KINTO_ADDONS_CHECKED_SECONDS`) ?
- Should we handle the case where certs are managed via XML and addon/plugins via Kinto? → only one pref for everything ?

Later with mark:
- [ ] Plug call to client `maybeSync()` with polling (see https://gist.github.com/mozmark/4b85d429c627bab7cc25)
- [ ] Load initial dump into SQLite https://bugzilla.mozilla.org/show_bug.cgi?id=1224528

**Comment by leplatrem** [comment]
Push to try submitted: https://treeherder.mozilla.org/#/jobs?repo=try&revision=57406885babf

**Comment by leplatrem** [comment]
Current status:

This PR completely bypasses `_loadBlocklistFromXMLString` when going via Kinto.
Since this method has a lot of (intended) side-effects, we should make sure that loading JSON has the same consequences: ex send message to `GFXInfoBase`

Question: should we handle the case where certs are managed via XML and addon/plugins via Kinto?

File: toolkit/mozapps/extensions/nsBlocklistService.js [hunk 3]
```diff
@@ -819,71 +838,140 @@ Blocklist.prototype = {
...
+      LOG("Blocklist::_preloadBlocklistFile: blocklist is disabled");
       return;
-    } catch (e) {
-      LOG("Blocklist::_preloadBlocklist: Failed to load XML file " + e)
     }
 
-    LOG("Blocklist::_preloadBlocklist: no XML File found");
-  }),
+    const paths = gBlocklistFromXML ? [FILE_BLOCKLIST]
+                                    : ['blocklist-addons.json', 'blocklist-plugins.json'];
```
Comment (leplatrem) [inline toolkit/mozapps/extensions/nsBlocklistService.js]: missing gfx here

File: toolkit/mozapps/extensions/nsBlocklistService.js [hunk 4]
```diff
@@ -515,6 +537,9 @@ Blocklist.prototype = {
     if (!gBlocklistEnabled)
       return;
 
+    if (!gBlocklistFromXML)
+      return;
```
Comment (leplatrem) [inline toolkit/mozapps/extensions/nsBlocklistService.js:546]: We should still download the XML if the Cert relies on it (cf question about keeping a pref per blocklist)

File: toolkit/mozapps/extensions/nsBlocklistService.js [hunk 5]
```diff
@@ -1218,6 +1505,13 @@ Blocklist.prototype = {
     return blockEntry.infoURL;
   },
 
+  _notifyObserversBlocklistGFX: function () {
+    // Notify `GfxInfoBase`, by passing a JSON serialization.
+    // This way we avoid spreading XML structure logics everywhere.
+    const payload = JSON.stringify(this._gfxEntries);
+    Services.obs.notifyObservers(null, "blocklist-data-gfxItems", payload);
```
Comment (leplatrem) [inline toolkit/mozapps/extensions/nsBlocklistService.js]: This is WIP: the observer still expects some XML!

**Comment by leplatrem** [comment]
This branch was split into several ones to ease review:
- #3 
- #4 
- #5

**Comment by leplatrem** [comment]
Closing in favor of smaller-distinct branches

**Pull request closed without merging**